### PR TITLE
jackass: init at v0.2.3

### PR DESCRIPTION
**Pull request by PowerUser64**
###### Description of changes

This PR adds [falkTX/jackass](https://github.com/falkTX/JackAss).

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/audio/jackass/default.nix [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+      url = "https://web.archive.org/web/20181016150224if_/https://download.steinberg.net/sdk_downloads/${name}.zip";
+      sha256 = "0da16iwac590wphz2sm5afrfj42jrsnkr1bxcy93lj7a369ildkj";
+    };
+    installPhase = ''
+      cp -r . $out
+    '';
+  };
+
+in
+stdenv.mkDerivation rec {
```
Comment (drupol) [inline pkgs/applications/audio/jackass/default.nix:17]: I think we should use the `finalAttrs` pattern here.

Let me share a couple of useful links regarding this:

1. History: https://discourse.nixos.org/t/avoid-rec-expresions-in-nixpkgs/8293
2. Documentation: https://nixos.org/manual/nixpkgs/unstable/#mkderivation-recursive-attributes
3. Recent example of implementation: https://github.com/NixOS/nixpkgs/compare/17f96f7b978e61576cfe16136eb418f74fab9952..9e6ea843e473d34d4f379b8b0d8ef0425a06defe

File: pkgs/applications/audio/jackass/default.nix [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+  pname = "jackass";
+  version = "1.1";
+
+  src = fetchFromGitHub {
+    owner = "falkTX";
+    repo = "JackAss";
+    rev = "v${version}";
+    sha256 = "sha256-6mqG4H6iGvDbGnmMeP/vcvSnvUGClZUl06XpKovt50E=";
```
Comment (drupol) [inline pkgs/applications/audio/jackass/default.nix:25]:
```suggestion
    hash = "sha256-6mqG4H6iGvDbGnmMeP/vcvSnvUGClZUl06XpKovt50E=";
```

File: pkgs/applications/audio/jackass/default.nix [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+  patchPhase = ''
+    cp -r ${vst-sdk}/VST2_SDK/{public.sdk,pluginterfaces} vstsdk2.4
+  '';
+
+  nativeBuildInputs = [ pkg-config ] ++ lib.optionals enableJackAssWine64 [ wine64 ];
+
+  makeFlags = [ "linux" ] ++ lib.optionals enableJackAssWine64 [ "wine64" ];
+
+  installPhase = ''
```
Comment (drupol) [inline pkgs/applications/audio/jackass/default.nix:36]: Missing hooks in `installPhase`.

**Review by drupol — CHANGES_REQUESTED** [review]

**Comment by PowerUser64** [comment]
Note, I changed the meta to have multiple licenses because vst2 sdk is unfree.

**Review by drupol — CHANGES_REQUESTED** [review]
Made a few more comments. Basically the same remarks but for the secondary derivation.

Please, let's be consistent when writing these even if using the finalAttrs pattern does not make a lot of difference there.

The commit log should be: `jackass: init at 0.2.3`

Thanks.

**Comment by PowerUser64** [comment]
Honestly we should probably just package the vst2sdk on its own, it's used in quite a few places.

**Comment by drupol** [comment]
Thanks.

Are you going to create a new PR or update this one ?

**Comment by PowerUser64** [comment]
Sorry, I've been away for a bit. I think I'll make it a separate PR. In it, I'll change this and all the other packages to use it.

**Comment by drupol** [comment]
Fair enough thanks!

**Comment by PowerUser64** [comment]
Could we merge this in its current state and then go through and make the change with the vst2 sdk afterwards? It works in its current state and it would be easiest to merge this and then add the vst2 sdk in another PR.

**Comment by drupol** [comment]
Why not make things correctly right now so we avoid postponing and avoid introducing things that need to be done at a later moment ?

**Comment by PowerUser64** [comment]
Everything with the vst2 sdk is made incorrectly right now, and I think conforming with the precedent would be fine for one last package. After that, we make a PR to make them all correct. That way we have one PR for each of the changes and we're not adding two packages in this one PR.

**Comment by PowerUser64** [comment]
Whew, it's been a while, but I'm able to work on this now.

I've realized we probably shouldn't package the vst2 sdk. The reason is that the maker of vst2, Steinberg, has discontinued it, and part of the vst2 license agreement states that you can't use it to make new things or distribute it after they discontinue it. Adding it as a package to nixpkgs feels like it's in even further violation of this. It's not used in many places, and usage of the vst2 sdk has died down significantly, but the terms are still there and it's clear Steinberg doesn't want people to keep using it.

Steinberg has graciously turned a blind eye to the fact that archive.org has been hosting a copy of it that a number of linux distros (including us) have been relying on to build stuff. That or they actually don't know about it. Either way, it is still technically not legal for archive.org to distribute this. Adding a package for the vst2 sdk would make it easy to include in development environments and perpetuate this, which could draw attention to the existence of the copy on archive.org and lead to its removal.

Anyway, long story short, I think adding the vst2 sdk as a package would be ill-guided and could risk losing us access to the vst2 sdk in the first place, so I really think we shouldn't package it at all.

**Comment by tobiasBora** [comment]
The licensing of vst2 is indeed tricky, we already discussed it here https://github.com/NixOS/nixpkgs/issues/145607 and ended up creating two versions of bespokesynth, one compiled without the vst2 and one with that is not prebuilt by hydra.

**Review by drupol — CHANGES_REQUESTED** [review]
Please move the package in `pkgs/by-name` directory.

Also, don't post your PR on discourse if it is in draft .

**Comment by PowerUser64** [comment]
Oh, I didn't know I was able to change it from draft. I'll do that.

> Please move the package in pkgs/by-name directory.

👍 will do

File: pkgs/by-name/ja/jackass/package.nix [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+  version = "1.1";
+
+  src = fetchFromGitHub {
+    owner = "falkTX";
+    repo = "JackAss";
+    rev = "v${finalAttrs.version}";
+    hash = "sha256-6mqG4H6iGvDbGnmMeP/vcvSnvUGClZUl06XpKovt50E=";
+  };
+
+  patchPhase = ''
```
Comment (drupol) [inline pkgs/by-name/ja/jackass/package.nix:30]: How about just using `postPatch` instead?

**Review by drupol — CHANGES_REQUESTED** [review]

**Comment by PowerUser64** [comment]
> @PowerUser64 is it working also without the proprietary vst? If so, you can maybe create two packages like we do for bespokesynth, one fully open source and one that is not.

That is an interesting question, it reminds me of https://github.com/Xaymar/vst2sdk, maybe I should ask falkTX (maintainer of jackass) if it could be written to include it instead of the proprietary vst sdk. However I'm guessing that's not what you were asking about. Unfortunately there aren't any versions of this plugin that don't use the proprietary vst sdk such as some other plugins do (like an lv2, vst3, or clap version). If there were other versions I'd say we should do that. I'll mention this to falk.

**Comment by tobiasBora** [comment]
This is very interesting, I wasn't aware of its existence. Is it stable? I was just refering to the fact that if jackass also comes with vst3 or lv, we can optionally enable vst2 if the user is ready to use non-free softwares.

**Comment by PowerUser64** [comment]
As far as I know, it's stable. It is being used in https://github.com/DISTRHO/DPF and there have only been 4 commits to the file since it was added ([the file is here](https://github.com/DISTRHO/DPF/blob/1504e7d327bfe0eac6a889cecd199c963d35532f/distrho/src/xaymar-vst2/vst.h)). It would be nice if maybe we could replace the vst2sdk here (or in bespoke) with this, although I'm not sure if we should because that seems like deviating pretty far from the original software.
cc @falkTX

**Comment by falkTX** [comment]
the way forward is porting the plugin to DPF. the only missing thing on DPF side is a way to fetch the host name size, which the jackass plugin uses for naming the jack client.

File: pkgs/by-name/ja/jackass/package.nix (hunk 4 above)
Comment (PowerUser64) [reply pkgs/by-name/ja/jackass/package.nix]: Done.

**Review by drupol — APPROVED** [review]

**Comment by PowerUser64** [comment]
Thanks for the review @drupol!
